firebase-config/src/main/java/com/google/firebase/remoteconfig/internal/ConfigRealtimeHttpClient.java

```diff
@@ -205,24 +223,60 @@ private void setRequestParams(HttpURLConnection httpURLConnection) throws IOExce
     outputStream.close();
   }
 
-  private synchronized void resetRetryParameters() {
-    httpRetrySeconds = random.nextInt(5) + 1;
-    httpRetriesRemaining = ORIGINAL_RETRIES;
-  }
-
   private synchronized void propagateErrors(FirebaseRemoteConfigException exception) {
     for (ConfigUpdateListener listener : listeners) {
       listener.onError(exception);
     }
   }
 
-  private synchronized void enableBackoff() {
-    this.isRealtimeDisabled = true;
+  @SuppressLint("VisibleForTests")
+  public int getNumberOfFailedStream() {
+    return metadataClient.getRealtimeBackoffMetadata().getNumFailedStreams();
+  }
+
+  @SuppressLint("VisibleForTests")
+  public Date getBackoffDate() {
+    return metadataClient.getRealtimeBackoffMetadata().getBackoffEndTime();
+  }
+
+  // TODO(issues/265): Make this an atomic operation within the Metadata class to avoid possible
+  // concurrency issues.
+  /**
+   * Increment the number of failed stream attempts, increase the backoff duration, set the backoff
+   * end time to "backoff duration" after {@code lastFailedRealtimeStreamTime} and persist the new
+   * values to disk-backed metadata.
+   */
+  private void updateBackoffMetadataWithLastFailedStreamConnectionTime(
+      Date lastFailedRealtimeStreamTime) {
+    int numFailedStreams = metadataClient.getRealtimeBackoffMetadata().getNumFailedStreams();
+
+    numFailedStreams++;
+
+    long backoffDurationInMillis = getRandomizedBackoffDurationInMillis(numFailedStreams);
+    Date backoffEndTime =
+        new Date(lastFailedRealtimeStreamTime.getTime() + backoffDurationInMillis);
+
+    metadataClient.setRealtimeBackoffMetadata(numFailedStreams, backoffEndTime);
+  }
+
+  /**
+   * Returns a random backoff duration from the range {@code timeoutDuration} +/- 50% of {@code
+   * timeoutDuration}, where {@code timeoutDuration = }{@link
+   * #BACKOFF_TIME_DURATIONS_IN_MINUTES}{@code [numFailedStreams-1]}.
+   */
+  private long getRandomizedBackoffDurationInMillis(int numFailedStreams) {
+    // Index of backoff duration. Use the minimum of numFailedFetch or the max length of backoff
+    // duration array.
+    int idx = Math.min(numFailedStreams, BACKOFF_TIME_DURATIONS_IN_MINUTES.length);
```

Comment:
I think this would be easier to read as a conditional. Thoughts?

```
int maxBackoffAttempts = BACKOFF_TIME_DURATIONS_IN_MINUTES.length;
long timeoutDuration = MINUTES.toMillis(BACKOFF_TIME_DURATIONS_IN_MINUTES[maxBackoffAttempts];

// Shorten the duration if there are remaining backoff attempts.
if (numFailedStreams < maxBackoffAttempts) {
  timeoutDuration = MINUTES.toMillis(BACKOFF_TIME_DURATIONS_IN_MINUTES[numFailedStreams])
}

return timeoutDuration / 2 + random.nextInt((int) timoutDuration);
```

Reply:
Yeah I like the conditional idea, I updated it to have a variation of this